engine/kv_engine.cpp

```diff
@@ -1229,51 +1229,89 @@ Status KVEngine::Modify(const StringView key, std::string* new_value,
   if (s != Status::Ok) {
     return s;
   }
+
   auto hint = hash_table_->GetHint(key);
   std::unique_lock<SpinMutex> ul(*hint.spin);
   version_controller_.HoldLocalSnapshot();
   defer(version_controller_.ReleaseLocalSnapshot());
   TimeStampType new_ts = version_controller_.GetLocalSnapshot().GetTimestamp();
-  auto ret = lookupKey<false>(key, static_cast<RecordType>(StringRecordType));
+  auto ret = lookupKey<true>(key, static_cast<RecordType>(StringRecordType));
+
+  StringRecord* existing_record = nullptr;
+  std::string existing_value;
+  std::string new_value;
   // push it into cleaner
-  if (ret.s == Status::Expired && ret.entry_ptr->IsTTLStatus()) {
-    hash_table_->UpdateEntryStatus(ret.entry_ptr, HashEntryStatus::Expired);
-    ul.unlock();
-    delayFree(OldDeleteRecord{ret.entry.GetIndex().ptr, ret.entry_ptr,
-                              PointerType::HashEntry, new_ts, hint.spin});
-    return Status::NotFound;
-  }
   if (ret.s == Status::Ok) {
-    StringRecord* old_record = ret.entry.GetIndex().string_record;
+    existing_record = ret.entry.GetIndex().string_record;
+    existing_value.assign(existing_record->Value().data(),
+                          existing_record->Value().size());
+  } else if (ret.s == Status::Outdated) {
+    existing_record = ret.entry.GetIndex().string_record;
+  } else if (ret.s == Status::NotFound) {
+    // nothing todo
+  } else {
+    return ret.s;
+  }
 
-    std::string modified_value = modify_func(old_record->Value(), modify_args);
-    if (!CheckValueSize(modified_value)) {
-      return Status::InvalidDataSize;
-    }
+  auto modify_operation =
+      modify_func(key, ret.s == Status::Ok ? &existing_value : nullptr,
+                  &new_value, modify_args);
+  switch (modify_operation) {
+    case ModifyOperation::Write: {
+      if (!CheckValueSize(new_value)) {
+        return Status::InvalidDataSize;
+      }
+      SpaceEntry space_entry =
+          pmem_allocator_->Allocate(StringRecord::RecordSize(key, new_value));
+      if (space_entry.size == 0) {
+        return Status::PmemOverflow;
+      }
 
-    auto requested_size =
-        key.size() + modified_value.size() + sizeof(StringRecord);
-    SpaceEntry space_entry = pmem_allocator_->Allocate(requested_size);
-    if (space_entry.size == 0) {
-      return Status::PmemOverflow;
+      StringRecord* new_record =
+          pmem_allocator_->offset2addr_checked<StringRecord>(
+              space_entry.offset);
+      StringRecord::PersistStringRecord(
+          new_record, space_entry.size, new_ts, StringDataRecord,
+          existing_record == nullptr
```

Comment:
if record is expired, don't need to execute modify write.

Reply:
We need, we treat it as deleted so we pass nullptr as old value to modify function